packages/ember-metal/lib/streams/stream.js

```diff
@@ -85,43 +110,51 @@ Stream.prototype = {
   },
 
   subscribe: function(callback, context) {
-    if (this.subscribers === undefined) {
-      this.subscribers = [callback, context];
+    var subscriber = new Subscriber(callback, context, this);
+    if (this.subscriberHead === null) {
+      this.subscriberHead = this.subscriberTail = subscriber;
     } else {
-      this.subscribers.push(callback, context);
+      var tail = this.subscriberTail;
+      tail.next = subscriber;
+      subscriber.prev = tail;
+      this.subscriberTail = subscriber;
     }
+
+    var stream = this;
+    return function() { subscriber.removeFrom(stream); };
```

Comment:
Ah, that makes sense. While the encapsulation is good, I am wary of the memory implications. I'll run it on ryan's dbmon demo tomorrow and EvilTrout's performance suite to see what the difference looks like for allocations.